torch_frame/gbdt/tuned_lightgbm.py

```diff
@@ -0,0 +1,201 @@
+
+import numpy as np
+import torch
+from torch import Tensor
+
+from torch_frame import DataFrame, Metric, TaskType, TensorFrame, stype
+from torch_frame.gbdt import GBDT
+
+
+class LightGBM(GBDT):
+    r"""A LightGBM model implementation with hyper-parameter tuning using
+    Optuna.
+
+    This implementation extends GBDT and aims to find optimal hyperparameters
+    by optimizing the given objective function.
+    """
+    def _to_lightgbm_input(
+        self,
+        tf: TensorFrame,
+    ) -> Tuple[np.ndarray, np.ndarray, List[str]]:
+        r"""Convert :class:`TensorFrame` into LightGBM-compatible input format:
+        :obj:`(feat, y)`.
+
+        Args:
+            tf (Tensor Frame): Input :obj:TensorFrame object.
+
+        Returns:
+            feat (numpy.ndarray): Output :obj:`numpy.ndarray` by
+                concatenating tensors of numerical and categorical features of
+                the input :class:`TensorFrame`.
+            y (numpy.ndarray): Prediction target :obj:`numpy.ndarray`.
+        """
+        tf = tf.cpu()
+        y = tf.y
+        assert y is not None
+
+        feats: List[Tensor] = []
+
+        if stype.categorical in tf.feat_dict:
+            feats.append(tf.feat_dict[stype.categorical])
+
+        if stype.numerical in tf.feat_dict:
+            feats.append(tf.feat_dict[stype.numerical])
+
+        if stype.text_embedded in tf.feat_dict:
+            feat = tf.feat_dict[stype.text_embedded]
+            feat = feat.view(feat.size(0), -1)
+            feats.append(feat)
+
+        # TODO Add support for other stypes.
+
+        if len(feats) == 0:
+            raise ValueError("The input TensorFrame object is empty.")
+
+        feat = torch.cat(feats, dim=-1)
+
+        return feat.numpy(), y.numpy()
+
+    def _predict_helper(
+        self,
+        model: Any,
+        x: DataFrame,
+    ) -> np.ndarray:
+        r"""A helper function that applies the lightgbm model on DataFrame
+        :obj:`x`.
+
+        Args:
+            model (lightgbm.Booster): The lightgbm model.
+            x (DataFrame): The input `DataFrame`.
+
+        Returns:
+            pred (np.ndarray): The prediction output.
+        """
+        pred = model.predict(x)
+        if self.task_type == TaskType.MULTICLASS_CLASSIFICATION:
+            pred = pred.argmax(axis=1)
+
+        return pred
+
+    def objective(
+        self,
+        trial: Any,
+        tf_train: TensorFrame,
+        tf_val: TensorFrame,
+        num_boost_round: int,
+    ):
+        r"""Objective function to be optimized.
+
+        Args:
+            trial (optuna.trial.Trial): Optuna trial object.
+            tf_train (TensorFrame): Train data.
+            tf_val (TensorFrame): Validation data.
+            num_boost_round (int): Number of boosting round.
+
+        Returns:
+            float: Best objective value. Mean absolute error for
+            regression task and accuracy for classification task.
+        """
+        import lightgbm
```

Comment:
You need to add the dependency of lightgbm in `pyproject.toml`

Reply:
Fixed.